File: Libraries/EntityRegistry/main/java/io/denery/entity/CustomizableEntity.java
package io.denery.entity;

import com.ticxo.modelengine.api.ModelEngineAPI;
import com.ticxo.modelengine.api.model.ActiveModel;
import com.ticxo.modelengine.api.model.ModeledEntity;
import io.denery.EntityRegistry;
import io.denery.behaviour.Behaviour;
import org.bukkit.entity.LivingEntity;
import org.bukkit.entity.Mob;

import java.util.Optional;

public abstract class CustomizableEntity {
    private final LivingEntity origin;
    private final Mob originMob;
    private final ActiveModel model;
    private final ModeledEntity modeledEntity;
    private final CustomizableEntityType type;
    private Behaviour behaviour;
    public CustomizableEntity(String registeredName) {
        this.type = EntityRegistry.getInstance().getRegisteredEntities().get(registeredName);
        this.origin = type.origin();
        this.originMob = (Mob) origin;
        this.behaviour = type.behaviour();

        Optional<ActiveModel> am =  Optional.ofNullable(ModelEngineAPI.api.getModelManager().createActiveModel(registeredName));
        if (am.isEmpty()) throw new RuntimeException("Couldn't load model: " + registeredName);
        model = am.get();

        Optional<ModeledEntity> me = Optional.ofNullable(ModelEngineAPI.api.getModelManager().createModeledEntity(origin));
        if (me.isEmpty()) throw new RuntimeException("Couldn't create modeled entity: " + origin);
        modeledEntity = me.get();

        modeledEntity.addActiveModel(model);
        modeledEntity.detectPlayers();
        modeledEntity.setInvisible(true);
    }

    public LivingEntity getOrigin() {
        return origin;
    }

    public CustomizableEntityType getType() {
        return type;
    }

    public ActiveModel getModel() {
        return model;
    }

    public ModeledEntity getModeledEntity() {
        return modeledEntity;
    }

    public Behaviour getBehaviour() {
        return behaviour;
    }

    public void setBehaviour(Behaviour behaviour) {
        this.behaviour = behaviour;
    }
}

File: apis/tablight-entity-registry-api-v0/src/main/mixed/io/denery/entityregistry/entity/CustomizableEntity.scala
package io.denery.entityregistry.entity

import com.ticxo.modelengine.api.ModelEngineAPI
import com.ticxo.modelengine.api.model.{ActiveModel, ModeledEntity}
import io.denery.entityregistry.EntityTypeRegistry
import io.denery.entityregistry.entity.CustomizableEntityType
import org.bukkit.entity.{LivingEntity, Entity}
import de.tr7zw.nbtapi.NBTEntity

import java.util.Optional
import scala.concurrent.ExecutionContext.Implicits.global

/**
 * Class which instance represents entity existing in the world that you can highly customize with model,
 * Behaviour (AI) and much more...
 *
 * @param _entityType type of entity which you registered before.
 * @param _origin the origin entity which custom entity is based. (Ocelot, villager, etc.)
 */
class CustomizableEntity(_entityType: CustomizableEntityType, _origin: LivingEntity):

  def modelEntity(): Unit = {
    val activeModel = ModelEngineAPI.api.getModelManager.createActiveModel(entityType.getID
      .getOrElse(throw RuntimeException("Please, set entity's ID to find model engine's model!")))
    val me = Optional.ofNullable(ModelEngineAPI.api.getModelManager.createModeledEntity(origin))
    if (me.isEmpty) throw new RuntimeException("Couldn't create modeled entity: " + origin.getName)
    val modeledEntity: ModeledEntity = me.get
    modeledEntity.addActiveModel(activeModel)
    modeledEntity.detectPlayers()
    modeledEntity.setInvisible(true)
  }

  def tag(): Unit = {
    val nbtent: NBTEntity = NBTEntity(_origin.asInstanceOf[Entity])
    //nbtent.setString()
  }

  def origin: LivingEntity = _origin

  def entityType: CustomizableEntityType = _entityType